tests/instruments/test_user_api.py

```diff
@@ -0,0 +1,120 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2021 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import pytest
+import serial
+
+from pymeasure.adapters import SerialAdapter, VISAAdapter
+from pymeasure.instruments import Instrument
+from pyvisa.constants import InterfaceType
+
+
+class MultiprotocolInstrument(Instrument):
+    """Test instrument for testing the configuration and use of multiple connection methods.
+
+    The instrument support multiple transports methods(serial, TCP/IP, GPIB).
+    In case of serial connections, the default baud rate is low (2400), but configurable in
+    the device, so it can use faster communication if desired.
+    Two examples in our instruments are AMI430, Keithley2260B (for TCP/IP, there are more
+    with GPIB capability)
+
+    It also has an attribute to configure a non-serial (gpib) connection. This demonstrates
+    that there may be attributes that are not valid for a serial connection, only others.
+    """
+    def __init__(self, adapter, name="Instrument with multiple connection methods", **kwargs):
+        # As an instrument contributor, I want to define default connection settings for my
+        # instrument with a minimum of boiler-plate. These settings should be easily
+        # user-overridable if appropriate without having to change the code/implementation
+        # of my instrument.
+
+        VISAAdapter.setdefaults(adapter, InterfaceType.asrl, kwargs,
+                                baud_rate = 2400)
+
+        VISAAdapter.setdefaults(adapter, InterfaceType.gpib, kwargs,
+                                enable_repeat_addressing = False)
+
+        super().__init__(adapter,
+                         name=name,
+                         **kwargs)
```

Comment:
See #539 for an implementation proposal.

Reply:
@bilderbuchi thanks a lot for your review. Let me recap the target of modifications also to double check my understanding is correct.
1. We are trying to define a mechanism to allow instruments developer's to easily define default values in presence of multiple interfaces and when these default values are different for each interface
2. The only case is `VISAAdapter` currently managed by `pyvisa`.
3. We need to apply the default only when user's provide instrument address as a number or string since these are the only cases where `Instrument` class creates a `VISAAdapter` class for you.
4. If user passes an adapter class as instrument address, no default should be set (and in any case all kwargs are ignored)

Consequences:

1. If the instrument has only one interface or the default apply to all interfaces then `kwargs.setdefault` should do the job, this is just python no special support.
2. I proposed to have an helper function to set the defaults for instrument developers, I do not see any use of it for regular users

3. Since this make sense only for `VISAAdapter`, I defined it as a static method of such class. This helper function needs to know:

- the instrument address to understand which will be the interface type associated
- the interface type for which the defaults needs to be applied, this is just a number.
- the user provided keywords
- the default definitions via keyword parameters

Again this is an helper function and not a proper method of `VISAAdapter`, so it is meant to be used outside a class instance and actually it has no use for class instances.
This helper function is, in my opinion, more than adequate to handle corner cases of instruments with multiple interfaces which needs specific defaults on some of them. It could handle also all the `pyvisa` interfaces and not just `aslr` and `gpib`

You proposed method will require to add a specific parameter for each interface supported by `VISAAdapter` which in turn are the interfaces supported by `pyvisa`.

Finally, let me share my personal preference regarding the possible solution to this problem (in order of priority):
1. Just remove safeKeywords and nothing else. We are left with marginal problems that can be solved by users by definining their own adapter instances. For example, as it is today the Fluke7341 will fail, if a GPIB address is used. I know how to fix this, but it is a bit clunky.
2. Use the helper function  `VISAAdapter.setdefaults`, same as 1. but more elegant.
3. Add 10 parameters to `Instrument.__init__` to handle currently supported pyvisa interface `[ <InterfaceType.gpib: 1>, <InterfaceType.vxi: 2>, <InterfaceType.gpib_vxi: 3>, <InterfaceType.asrl: 4>, <InterfaceType.pxi: 5>, <InterfaceType.tcpip: 6>, <InterfaceType.usb: 7>, <InterfaceType.rio: 8>, <InterfaceType.firewire: 9>, <InterfaceType.rsnrp: 33024>]`. Obviously we can safely assume that no instrument offer all of them simultaneously, some interfaces I have no idea of what they are and they could be filtered out, but again why taking decision for the users which offer only risks, i.e. preventing to define defaults for the `rio` (whatever it is) interface, just to discover that we can create inconvenience to some users (similar issues created by `safeKeyword` approach).

Let me say also that I like very much the review process because offer the possibility to share valuable opinion between people with different background and culture.

Sorry for the long comment and thanks for your patience if you read up to here.

Reply:
1-4: I agree with your understanding
Consequences: Agreed, this was clear to me from your code & comments :-) 

I think i see a "better"/cleaner way to reach these goals which is what I propose in #538 (cue Raymond Hettinger)
![image](https://user-images.githubusercontent.com/327442/145201238-195202fb-1d44-46e5-bd65-8d59c8e0c17c.png)

> You proposed method will require to add a specific parameter for each interface supported by VISAAdapter which in turn are the interfaces supported by pyvisa.

To clarify: This is only needed for any interface-specific args of the instrument. So, 
* if there are only generally valid ones, they go into  **kwargs as before.
* If the instrument only has a serial interface, use **kwargs as before
* If there is _another_ interface, simply move the asrl-specific ones into a dict ` asrl={}`
* There is no need to stub out or define the whole set of visa interfaces, the code checks against the enum of possible choices and removes invalid ones as appropriate.

Preferences:
Now I'm not sure if you have seen #538 yet, maybe your comment writing overlapped my push. 
In any case, I'm fully with you re: removing safekewords. It is broken, it has to go. While we are doing this, I want to offer a flexible and fully-featured solution for the future, to avoid limitations where feasible.

I think with #538 I am offering an improved 3.b, usable for all pyvisa interfaces. I don't think the "pollution" of Instrument.__init__ with 10 more parameters is too bad because they are all optional, and can (and should) be documented in one go in an entry for `**kwargs` (that this is missing should also be fixed, we should document what we expect in **kwargs).

> Let me say also that I like very much the review process because offer the possibility to share valuable opinion between people with different background and culture.

Thank you, that is a relief to me. Text-based communication is so limited in some aspects. Sometimes I'm not sure if my mostly corrective/negative feedback is not irritating to you, as we seem to often "pull" in different directions regarding coding philosophy, and I try to be cautious with any changes to minimize future maintenance burden.

Reply:
> Thank you, that is a relief to me. Text-based communication is so limited in some aspects. Sometimes I'm not sure if my mostly corrective/negative feedback is not irritating to you, as we seem to often "pull" in different directions regarding coding philosophy, and I try to be cautious with any changes to minimize future maintenance burden.

I think that feedback positive or negative is always useful, indeed l always experienced good outcome from discussion with other people.
I also appreciate gut feeling feedback that normally could detect problems that with normal review are not surfacing.

This does not mean that we are always in agreement and indeed we have different coding philosophy, but let's try to come to a conclusion. Possibly having some other people in the discussion could help.

I will have a look at your PR #539 and get back to you.